File: frontend/transactions.js
import apiClient from "./client";

export const listTransactions = async () => {
  const { data } = await apiClient.get("/transactions");
  return data;
};

export const createTransaction = async ({ date, description, type, amount }) => {
  const { data } = await apiClient.post("/transactions", {
    date,
    description,
    type,
    amount,
  });
  return data;
};

export const updateTransaction = async (id, { date, description, type, amount }) => {
  const { data } = await apiClient.put(`/transactions/${id}`, {
    date,
    description,
    type,
    amount,
  });
  return data;
};

export const deleteTransaction = async (id) => {
  await apiClient.delete(`/transactions/${id}`);
};

File: transactions.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.core.database import get_db
from app.models.user import User
from app.schemas.transaction import TransactionCreate, TransactionOut, TransactionUpdate
from app.services.transaction_service import TransactionService

router = APIRouter(prefix="/transactions", tags=["transactions"])


@router.get("", response_model=list[TransactionOut])
def list_transactions(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = TransactionService(db)
    return service.list_transactions(current_user.id)


@router.post("", response_model=TransactionOut, status_code=status.HTTP_201_CREATED)
def create_transaction(
    data: TransactionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = TransactionService(db)
    return service.create_transaction(current_user.id, data)


@router.get("/{transaction_id}", response_model=TransactionOut)
def get_transaction(
    transaction_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = TransactionService(db)
    return service.get_transaction(transaction_id, current_user.id)


@router.put("/{transaction_id}", response_model=TransactionOut)
def update_transaction(
    transaction_id: int,
    data: TransactionUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = TransactionService(db)
    return service.update_transaction(transaction_id, current_user.id, data)


@router.delete("/{transaction_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_transaction(
    transaction_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = TransactionService(db)
    service.delete_transaction(transaction_id, current_user.id)
    return None
